File: src/ai/hazardSpread.ts
import { HazardZone, GraphNode, GraphEdge, DisasterType } from '../types/graph';
import { WeatherCondition } from '../types/simulation';

// Euclidean distance helper in km
function distanceKm(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const dLat = (lat2 - lat1) * 111.32;
  const dLng = (lng2 - lng1) * 111.32 * Math.cos((lat1 * Math.PI) / 180);
  return Math.sqrt(dLat * dLat + dLng * dLng);
}

// Compute hazard intensity at a given coordinate
export function computePointHazardRisk(
  lat: number,
  lng: number,
  elevation: number,
  hazards: HazardZone[],
  weather: WeatherCondition
): number {
  let maxRisk = 0;

  for (const hazard of hazards) {
    const dist = distanceKm(hazard.centerLat, hazard.centerLng, lat, lng);

    switch (hazard.type) {
      case 'flood': {
        // Floods spread wider at low elevations and high rainfall
        const effectiveRadius = hazard.radiusKm * (1 + (weather.rainfallMmH / 50) * 0.5);
        const waterHeight = (hazard.waterLevelMeters || 3) - elevation * 0.1;
        if (dist <= effectiveRadius) {
          const proximity = 1 - dist / effectiveRadius;
          const floodRisk = proximity * hazard.intensity * (waterHeight > 0 ? 1.2 : 0.6);
          maxRisk = Math.max(maxRisk, Math.min(1.0, floodRisk));
        }
        break;
      }

      case 'wildfire': {
        // Wildfire expands elliptically in the direction of the wind
        const windRad = ((hazard.windAngleDeg ?? weather.windDirectionDeg) * Math.PI) / 180;
        const dx = (lng - hazard.centerLng) * 111.32 * Math.cos((lat * Math.PI) / 180);
        const dy = (lat - hazard.centerLat) * 111.32;

        // Project coordinate along wind vector
        const downwindDist = dx * Math.cos(windRad) + dy * Math.sin(windRad);
        const crosswindDist = -dx * Math.sin(windRad) + dy * Math.cos(windRad);

        const windSpeed = hazard.windSpeedKmH ?? weather.windSpeedKmH;
        const forwardRadius = hazard.radiusKm * (1 + windSpeed / 30);
        const flankRadius = hazard.radiusKm * 0.7;

        const ellipticalDistSq =
          Math.pow(Math.max(0, downwindDist) / forwardRadius, 2) +
          Math.pow(Math.min(0, downwindDist) / (hazard.radiusKm * 0.5), 2) +
          Math.pow(crosswindDist / flankRadius, 2);

        if (ellipticalDistSq <= 1.0) {
          const fireIntensity = (1 - Math.sqrt(ellipticalDistSq)) * hazard.intensity * (1 + (100 - weather.humidityPct) / 100 * 0.3);
          maxRisk = Math.max(maxRisk, Math.min(1.0, fireIntensity));
        }
        break;
      }

      case 'earthquake': {
        // Seismic intensity attenuates exponentially with distance and depth
        const depth = hazard.epicenterDepthKm || 10;
        const hypoDist = Math.sqrt(dist * dist + depth * depth);
        const seismicIntensity = (hazard.intensity * 15) / (hypoDist * 1.5 + 2);
        maxRisk = Math.max(maxRisk, Math.min(1.0, seismicIntensity));
        break;
      }

      case 'landslide': {
        // High risk if close to center and steep slope
        if (dist <= hazard.radiusKm) {
          const proximity = 1 - dist / hazard.radiusKm;
          const rainFactor = Math.min(1.5, 1 + weather.rainfallMmH / 40);
          maxRisk = Math.max(maxRisk, Math.min(1.0, proximity * hazard.intensity * rainFactor));
        }
        break;
      }

      case 'cyclone': {
        // Cyclonic eye and high wind speed radius
        if (dist <= hazard.radiusKm * 1.8) {
          const cycloneRisk = (1 - dist / (hazard.radiusKm * 1.8)) * hazard.intensity;
          maxRisk = Math.max(maxRisk, Math.min(1.0, cycloneRisk));
        }
        break;
      }
    }
  }

  return Math.round(maxRisk * 1000) / 1000;
}

// Advance hazard positions and sizes over simulation time delta (hours)
export function advanceHazards(
  hazards: HazardZone[],
  dtHours: number,
  weather: WeatherCondition
): HazardZone[] {
  return hazards.map(h => {
    let growthRate = h.expansionRateKmH;

    if (h.type === 'wildfire') {
      growthRate *= (1 + weather.windSpeedKmH / 40) * (1 - weather.humidityPct / 200);
      // Drift center downwind
      const windRad = (weather.windDirectionDeg * Math.PI) / 180;
      const driftSpeedKmH = weather.windSpeedKmH * 0.08;
      const dLat = (driftSpeedKmH * dtHours * Math.sin(windRad)) / 111.32;
      const dLng = (driftSpeedKmH * dtHours * Math.cos(windRad)) / (111.32 * Math.cos((h.centerLat * Math.PI) / 180));
      return {
        ...h,
        centerLat: h.centerLat + dLat,
        centerLng: h.centerLng + dLng,
        radiusKm: Math.min(15, h.radiusKm + growthRate * dtHours),
        intensity: Math.min(1.0, h.intensity + 0.02 * dtHours)
      };
    } else if (h.type === 'flood') {
      growthRate *= (1 + weather.rainfallMmH / 30);
      return {
        ...h,
        radiusKm: Math.min(18, h.radiusKm + growthRate * dtHours),
        waterLevelMeters: (h.waterLevelMeters || 3) + (weather.rainfallMmH / 100) * dtHours,
        intensity: Math.min(1.0, h.intensity + 0.03 * dtHours)
      };
    } else {
      return {
        ...h,
        radiusKm: Math.min(20, h.radiusKm + growthRate * dtHours)
      };
    }
  });
}

File: backend/ai/hazard_spread.py
import math
from typing import List
from ..models.graph import HazardZone
from ..models.simulation import WeatherCondition

def distance_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    d_lat = (lat2 - lat1) * 111.32
    d_lng = (lng2 - lng1) * 111.32 * math.cos(math.radians(lat1))
    return math.sqrt(d_lat * d_lat + d_lng * d_lng)

def compute_point_hazard_risk(
    lat: float,
    lng: float,
    elevation: float,
    hazards: List[HazardZone],
    weather: WeatherCondition
) -> float:
    max_risk = 0.0

    for hazard in hazards:
        dist = distance_km(hazard.centerLat, hazard.centerLng, lat, lng)

        if hazard.type == 'flood':
            effective_radius = hazard.radiusKm * (1.0 + (weather.rainfallMmH / 50.0) * 0.5)
            water_height = (hazard.waterLevelMeters or 3.0) - elevation * 0.1
            if dist <= effective_radius:
                proximity = 1.0 - dist / effective_radius
                flood_risk = proximity * hazard.intensity * (1.2 if water_height > 0 else 0.6)
                max_risk = max(max_risk, min(1.0, flood_risk))

        elif hazard.type == 'wildfire':
            wind_rad = math.radians(hazard.windAngleDeg if hazard.windAngleDeg is not None else weather.windDirectionDeg)
            dx = (lng - hazard.centerLng) * 111.32 * math.cos(math.radians(lat))
            dy = (lat - hazard.centerLat) * 111.32

            downwind_dist = dx * math.cos(wind_rad) + dy * math.sin(wind_rad)
            crosswind_dist = -dx * math.sin(wind_rad) + dy * math.cos(wind_rad)

            wind_speed = hazard.windSpeedKmH if hazard.windSpeedKmH is not None else weather.windSpeedKmH
            forward_radius = hazard.radiusKm * (1.0 + wind_speed / 30.0)
            flank_radius = hazard.radiusKm * 0.7

            fw_term = math.pow(max(0.0, downwind_dist) / forward_radius, 2)
            bw_term = math.pow(min(0.0, downwind_dist) / (hazard.radiusKm * 0.5), 2)
            cw_term = math.pow(crosswind_dist / flank_radius, 2)
            elliptical_dist_sq = fw_term + bw_term + cw_term

            if elliptical_dist_sq <= 1.0:
                fire_intensity = (1.0 - math.sqrt(elliptical_dist_sq)) * hazard.intensity * (1.0 + (100.0 - weather.humidityPct) / 100.0 * 0.3)
                max_risk = max(max_risk, min(1.0, fire_intensity))

        elif hazard.type == 'earthquake':
            depth = hazard.epicenterDepthKm or 10.0
            hypo_dist = math.sqrt(dist * dist + depth * depth)
            seismic_intensity = (hazard.intensity * 15.0) / (hypo_dist * 1.5 + 2.0)
            max_risk = max(max_risk, min(1.0, seismic_intensity))

        elif hazard.type == 'landslide':
            if dist <= hazard.radiusKm:
                proximity = 1.0 - dist / hazard.radiusKm
                rain_factor = min(1.5, 1.0 + weather.rainfallMmH / 40.0)
                max_risk = max(max_risk, min(1.0, proximity * hazard.intensity * rain_factor))

        elif hazard.type == 'cyclone':
            if dist <= hazard.radiusKm * 1.8:
                cyclone_risk = (1.0 - dist / (hazard.radiusKm * 1.8)) * hazard.intensity
                max_risk = max(max_risk, min(1.0, cyclone_risk))

    return round(max_risk, 3)

def advance_hazards(
    hazards: List[HazardZone],
    dt_hours: float,
    weather: WeatherCondition
) -> List[HazardZone]:
    updated_hazards = []
    for h in hazards:
        growth_rate = h.expansionRateKmH

        if h.type == 'wildfire':
            growth_rate *= (1.0 + weather.windSpeedKmH / 40.0) * (1.0 - weather.humidityPct / 200.0)
            wind_rad = math.radians(weather.windDirectionDeg)
            drift_speed = weather.windSpeedKmH * 0.08
            d_lat = (drift_speed * dt_hours * math.sin(wind_rad)) / 111.32
            d_lng = (drift_speed * dt_hours * math.cos(wind_rad)) / (111.32 * math.cos(math.radians(h.centerLat)))
            updated_hazards.append(h.model_copy(update={
                'centerLat': h.centerLat + d_lat,
                'centerLng': h.centerLng + d_lng,
                'radiusKm': min(15.0, h.radiusKm + growth_rate * dt_hours),
                'intensity': min(1.0, h.intensity + 0.02 * dt_hours)
            }))
        elif h.type == 'flood':
            growth_rate *= (1.0 + weather.rainfallMmH / 30.0)
            cur_water = h.waterLevelMeters or 3.0
            updated_hazards.append(h.model_copy(update={
                'radiusKm': min(18.0, h.radiusKm + growth_rate * dt_hours),
                'waterLevelMeters': cur_water + (weather.rainfallMmH / 100.0) * dt_hours,
                'intensity': min(1.0, h.intensity + 0.03 * dt_hours)
            }))
        else:
            updated_hazards.append(h.model_copy(update={
                'radiusKm': min(20.0, h.radiusKm + growth_rate * dt_hours)
            }))

    return updated_hazards
